### PR TITLE
Provide an option to optimize either wall clock time of replication or memory

### PR DESCRIPTION
**Pull request by schomatis**
Based on https://github.com/filecoin-project/rust-fil-proofs/pull/567.

Closes https://github.com/filecoin-project/rust-fil-proofs/issues/647.

### CODE REVIEW

**Comment by schomatis** [comment]
Without the intermediate anonymous `mmap` (for the sequential MT generation check) the max RSS for a sector size of 8 GiB is reduced (as expected) from 40 to 32 GiB, 4 sector sizes, the replication process (`transform_and_replicate_layers` ) thus takes now (the intended) 2 sector sizes (1 MT), the `zigzag` example retains one sector size in memory during the entire run (to run the extraction at the end) and there remains a sector size to be accounted for (that would add up to the measured 32 GiB).

**Comment by schomatis** [comment]
Tagging this for review but #567 should be reviewed and merged first.

**Comment by schomatis** [comment]
Trying to investigate the extra sector size in the replication process (see previous [comment](https://github.com/filecoin-project/rust-fil-proofs/pull/642#issuecomment-491514472)) I'm `strace`ing the `zigzag` example and observing a pattern of anonymous `mmap`s that I'm not identifying as part of the (new) disk-backed MT generation setup. Additionally, this pattern appears only with big sector sizes and starts out with the sector size and then successively reduces the size by a factor of 1/2. I'm _speculating_ then that this is a Rust allocation mechanism and that we're still allocating local space for the MT or (most probably) the replication `data` from where it is generated. I'll try to do an in-depth profile to track its source.

The pattern observed is, from a sector size of 256 MiB:

```
mmap(NULL, 268435456, PROT_READ|PROT_WRITE, MAP_SHARED, 5, 0) = 0x7f17e8000000  # File-backed `mmap` of the leaves (one sector size)
openat(AT_FDCWD, "/home/snarky/space/rust-fil-proofs-trees-dir/tree-1-3366915306-top-half", O_RDWR|O_CREAT|O_CLOEXEC, 0666) = 6
ftruncate(6, 268435456)                 = 0
fstat(6, {st_mode=S_IFREG|0664, st_size=268435456, ...}) = 0
mmap(NULL, 268435456, PROT_READ|PROT_WRITE, MAP_SHARED, 6, 0) = 0x7f17d8000000  # File-backed `mmap` of the top half of the MT (second sector size)

mmap(NULL, 268439552, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f17b7fff000  # Anonymous mmap of one sector size (presumed
                                                                                                   to be thread stack storage).
mmap(NULL, 268439552, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f1787fff000  # Another anonymous, same size.
munmap(0x7f17b7fff000, 268439552)       = 0                                                     # Unmapping first anonymous.
mmap(NULL, 134221824, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f17cffff000  # Mapping _half_ sector size now.
futex(0x55d575765418, FUTEX_WAKE_PRIVATE, 2147483647) = 8
futex(0x55d5757653c0, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x55d575773cb8, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x55d5757748e0, FUTEX_WAKE_PRIVATE, 1) = 0
munmap(0x7f1787fff000, 268439552)       = 0
munmap(0x7f17cffff000, 134221824)       = 0
mmap(NULL, 134221824, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f17cffff000
mmap(NULL, 134221824, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f17bffff000
munmap(0x7f17cffff000, 134221824)       = 0
mmap(NULL, 67112960, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f17d3fff000   # Now 1/4th of a sector size.
futex(0x55d57576541c, FUTEX_WAKE_PRIVATE, 2147483647) = 8
futex(0x55d5757653c0, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x55d575773cb8, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x55d5757748e0, FUTEX_WAKE_PRIVATE, 1) = 0
munmap(0x7f17bffff000, 134221824)       = 0
munmap(0x7f17d3fff000, 67112960)        = 0
mmap(NULL, 67112960, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f17d3fff000
mmap(NULL, 67112960, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f17cfffe000
munmap(0x7f17d3fff000, 67112960)        = 0
mmap(NULL, 33558528, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f17d5fff000  # 1/8th, and so on..
futex(0x55d575765418, FUTEX_WAKE_PRIVATE, 2147483647) = 8
futex(0x55d5757653c0, FUTEX_WAKE_PRIVATE, 1) = 0
```

The command run is:

```
export TREES_DIR=/home/snarky/space/rust-fil-proofs-trees-dir                &&
FIL_PROOFS_REPLICATED_TREES_DIR="$TREES_DIR"                                  \
RUST_BACKTRACE=1                                                              \
CARGO_INCREMENTAL=1                                                           \
nohup strace  cargo run                                       \
  -p filecoin-proofs                                                          \
  --release                                                                   \
  --example zigzag                                                            \
  --features                                                                  \
    disk-trees                                                                \
  --                                                                          \
    --size 262144                                                            \
    --no-bench                                                                \
    --hasher blake2s                                                          \
    --m 1  --expansion 0 --layers 1                                           \
     > zigzag-8gib.out &
```

(attaching the full `zigzag-8gib.out` which confusingly is named with the standard 8 GiB test size, this reduced 256 MiB size allows for faster tests preserving the allocation pattern)

/cc @porcuquine @dignifiedquire in case this pattern makes sense to you.

[zigzag-8gib.txt](https://github.com/filecoin-project/rust-fil-proofs/files/3170702/zigzag-8gib.txt)

**Comment by schomatis** [comment]
Also note that the current implementation in this PR does a _sequential_ MT generation-encoding so I also don't follow where are all the `futex` calls coming from.

**Comment by porcuquine** [comment]
> the `zigzag` example retains one sector size in memory during the entire run (to run the extraction at the end) and there remains a sector size to be accounted for (that would add up to the measured 32 GiB).

That extraction hasn't turned out to be very important and practice, and the memory usage is problematic for our benchmarks. Can we add it as command-line option defaulting to no-extraction/no-memory?

**Comment by schomatis** [comment]
> That extraction hasn't turned out to be very important and practice, and the memory usage is problematic for our benchmarks. Can we add it as command-line option defaulting to no-extraction/no-memory?

@porcuquine Actually my characterization of that memory region (`data`, as it's usually called) was too simplistic; it's not retained just for extraction purposes, it's (I think) heavily embedded in the entire replication logic (some functions even use it as scratch pad). I do agree that taking it out (and absorbing it as an MT `Store`) would be a very nice win but I'd have to take a closer look at it and I'll definitely need external support to carry it through (mentioning it just so you can evaluate the cost/benefit and the estimated time, we can discuss it in more depth in a separate issue).

**Comment by schomatis** [comment]
(Part of the motivation behind splitting between leaves and the rest of the MT was to provide storage options that would succeed the current overuse of `&mut [u8]` such as this one.)

File: storage-proofs/src/settings.rs [hunk 1]
```diff
@@ -23,6 +27,13 @@ impl Default for Settings {
             maximize_caching: false,
             merkle_tree_path: "/tmp/merkle-trees".into(),
             num_proving_threads: 1,
+            replicated_trees_dir: "".into(),
+            generate_merkle_trees_in_parallel: false,
+            // FIXME: Decide on a sensible default value, setting to `false`
+            //  at the moment for testing purposes only.
```
Comment (schomatis) [inline storage-proofs/src/settings.rs]: Need to decide this.

Comment (dignifiedquire) [reply storage-proofs/src/settings.rs]: which option are you referring to with `disk-trees`?

Comment (schomatis) [reply storage-proofs/src/settings.rs]: `generate_merkle_trees_in_parallel`

Comment (schomatis) [reply storage-proofs/src/settings.rs]: I suppose should be `true` as it's the current behavior, but wanted to make sure.

Comment (schomatis) [reply storage-proofs/src/settings.rs]: Misunderstood the original question I think, sorry, `disk-trees` is not an internal configuration option but the Rust feature that enables all the file-mapping related logic.

Comment (dignifiedquire) [reply storage-proofs/src/settings.rs]: right, you could set the default depending on the build flag

Comment (dignifiedquire) [reply storage-proofs/src/settings.rs]: You could put that logic into the config loading part, if you want

Comment (schomatis) [reply storage-proofs/src/settings.rs]: in `Settings.new`? should that happen before or after the `try_into`?

Comment (schomatis) [reply storage-proofs/src/settings.rs]: done

**Comment by dignifiedquire** [comment]
lets get this in, it has been long enough, any further improvements we can do in follow ups :)

**Pull request merged**